src/V3ReuseWideTemps.cpp

```diff
@@ -0,0 +1,246 @@
+//      For each temporary read operation:
+//          If temporary was recently replaced, find replacee and replace it.
+//          Replace dtype of a temporary with old one to use old width.
+//*************************************************************************
+
+#include "V3PchAstMT.h"
+
+#include "V3ReuseWideTemps.h"
+
+#include "V3Global.h"
+#include "V3Stats.h"
+#include "V3ThreadPool.h"
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+const string TEMP_PREFIX = "__Vtemp_";  // Prefix for temporary variables.
+
+class ReuseWideTempsVisitor final : public VNVisitor {
+    // TYPES
+    using WidthType = int;
+    using TempIdType = size_t;
+    using IdTempMap = std::map<TempIdType, AstVar*>;
+
+    // MEMBERS
+    AstCFunc* const m_cfuncp;  // Current function.
+
+    // STATE
+    std::map<WidthType, IdTempMap, std::greater<WidthType>>
+        m_freeTemps;  // Freed temporaries eligible for reuse.
+    std::map<WidthType, IdTempMap, std::greater<WidthType>>
+        m_replacedInStmt;  // Temporaries replaced in a current statement.
+    std::unordered_map<AstVar*, AstVar*> m_replaced;  // Mapping of vars replaced in the previous
+                                                      // WRITE operation to be used in READ.
+    TempIdType m_tmpVarCnt = 0;  // Counter for temporary ids.
+    std::unordered_set<AstVar*> m_removedTemps;  // No longer needed temporaries to be deleted.
+    VDouble0 m_totalTemps;  // Statistic for how many temps were before cleanup.
+
+    // STATIC METHODS
+    static TempIdType idFrom(const AstNode* nodep) {
+        UASSERT(VString::startsWith(nodep->name(), TEMP_PREFIX),
+                "Prefix must start with '" + TEMP_PREFIX + "'");
+        const auto it = nodep->name().find_last_of('_');
+        UASSERT(it != string::npos, "Unexpected temp name format: " + nodep->name());
+        const string idStr = nodep->name().substr(it + 1, nodep->name().size());
+        std::istringstream iss(idStr);
+        TempIdType id;
+        iss >> id;
+        UASSERT(iss, "Unexpected temp name format: " + nodep->name());
+        return id;
+    }
+    static bool isWideTemp(const AstVar* nodep) {
+        return nodep && nodep->isStatementTemp() && nodep->isWide()
+               && VN_IS(nodep->dtypep(), BasicDType);
+    }
+    static bool isExcluded(AstVar* varp) { return varp && varp->user1(); }
+    static void excludeVars(AstCFunc* nodep) {
+        nodep->foreach([](AstCCall* ccallp) {
+            for (AstNode* argsp = ccallp->argsp(); argsp; argsp = argsp->nextp()) {
+                if (AstVarRef* refp = VN_CAST(argsp, VarRef)) {
+                    if (refp->varp()) refp->varp()->user1SetOnce();
+                }
+            }
+        });
+    }
+    static void fixupType(AstVarRef* nodep, AstVar* varp) {
+        // Swap varp of current VarRef and preserve dtype to use old width.
+        AstNodeDType* prevDtypep = nodep->dtypep();
+        nodep->varp(varp);
+        nodep->dtypep(prevDtypep);
+    }
```

Comment:
We should really be moving towards a more sound type system for AST nodes, e.g.: where the type of a VarRef is precisely the type of the referenced variable. There are various places around the codebase where this assumption is made, and we should be fixing wherever it's violated, unfortunately this is a step in the opposite direction.